### DISCUSSION TITLE
distributing/overview

### DISCUSSION THREAD

**Discussion by giscus** [velopack/velopack.docs#17] [General]
# distributing/overview

Distributing with Velopack is extremely easy, it's usually just as simple as uploading your files somewhere that can be downloaded with HTTP. This means you could host them on an IIS or nodejs site, on shared file hosting such as AWS S3, Azure Storage, BackBlaze B2, or even for free on GitHub/GitLab releases if your project is open source.

https://docs.velopack.io/distributing/overview

**Comment by bearyung** [comment]
How's the UpdateManager.DownloadUpdatesAsync() determine download the full version or delta version? I've tried on Mac OS to build 1.0.0, upload that to online storage, setup the app and build a new 1.0.1 and upload that to the online storage.

I've found that the UpdateManager.DownloadUpdatesAsync() will download {app_name}-1.0.1-osx-full.nupkg instead of {app_name}-1.0.1-osx-delta.nupkg

**Reply by caesay** [reply]
Hello! Because of how the `.pkg` installer works, it cannot contain the files necessary for delta updates - so the first update on macos will do a full update instead. The second update and all others after that should use deltas.

**Reply by bearyung** [reply]
Thanks for the quick response! Have tried and confirm that the deltas works for 1.0.2 and so on!

**Comment by nabeelbaghoor** [comment]
Hey, I am on windows and am just distributing YourAppId-Portable.zip extracted as we don't need to install the app. The app is running and fetching the updates! Thanks. Now, in remote update folder, exactly which files i should upload in order to make updates work. Just these?
- releases.win
- full
- delta

or i need everything inside Releases

**Reply by caesay** [reply]
Only the files you've listed are required for updates.

**Comment by Mopy18** [comment]
Hey, I'm migrating from Clowd.Squirrel to Velopack and if I understand correctly of my next update I must upload both RELEASES and releases.{channel}.json files. 

What about the other updates that will come after? 
- Should I continue publishing both files until all my users have updated to Velopack? (but I have now way of knowing that)
- Is the releases.{channel}.json file enough? Like if a user has not opened the app for a long time and is still on the Clowd.Squirrel version, there will be 2 updates in a row? (one to update to the latest version available in RELEASES, and a second one that will use the releases.{channel}.json because now the app can use Velopack).

**Reply by caesay** [reply]
Your assumptions are correct. You don't need to keep updating the `RELEASES` file, but if you stop it means anyone migrating from Squirrel will get two updates (one from Squirrel to Velopack, and then another to the latest Velopack release). You will need to make that determination. It may be sensible to keep updating RELEASES for a short while, so your active users get a better update experience. Any users who have not been updated by that point will just get an extra update, not a big deal.

**Reply by Mopy18** [reply]
Great! Thank you for the fast answer, and thank you for developing Velopack you did an amazing job!

**Comment by rezaelahi-ui** [comment]
Hi,
How can I set an access token to download files from a protected folder on my Plesk website using Velopack?
I want to know how to send the token with Velopack.

**Reply by caesay** [reply]
This was answered here: https://github.com/velopack/velopack/issues/511

**Comment by benkoshy** [comment]
I learned the hard way - you cannot host on Google Drive. 

The URLs does not natrually resolve in the way you would expect it to.

**Comment by PavloNesterenkoVecUa** [comment]
I have created two releases and I see the "UpdatingToVersion" message, my application closes, and when it restarts the version is the same and it has not updated.

```
private async Task UpdateLocalAsync()
{
    var updateManager = new UpdateManager(...);
    var newVersion = await updateManager.CheckForUpdatesAsync();
    if (newVersion == null) { return; }
    MessageBox.Show("UpdatingToVersion");
    updateManager.ApplyUpdatesAndRestart(newVersion);
}
```

Also my goal is to allow the user to download only available versions from the server and save traffic. I dont wanna force every user to download all updates. What is minimal stack of files I copy from the release directory to give the user for updates?

What if user will try to skip some versions in update? For example user will have version 1.1.18 installed and latest version will be 1.1.20 . Is it will be necessary to download all versions to update?

**Reply by caesay** [reply]
Comments on documentation are intended for questions that relate to clarification of the docs themselves, and where the answers are likely to help others reading the docs. For help with any specific issue unique to you, please join our [Discord](https://discord.gg/CjrCrNzd3F) or open a GH issue if you think you've found a bug.

Your update process should consist of:
- `CheckForUpdatesAsync` (to see if there is an update available)
- `DownloadUpdatesAsync` (to download the required update files)
- `ApplyUpdatesAndRestart` (to install the downloaded update files)

Your release process should consist of:
- `dotnet publish` (to compile the new version of your application
- `vpk download` (to download the latest public version from wherever you host your updates, this is used to create deltas)
- `vpk pack` (to create a new full and delta update file)
- `vpk upload` (to distribute the releases to users by uploading the new update files, updating the remote configuration, and applying retention policies)
 
You should avoid copying releases out of the release folder manually as you are likely to get this wrong. In the event you are publishing to a fileshare or local folder, the `vpk upload local` command is available.

When updating, deltas are generally used - these are small and bandwidth efficient. The user will be required to download every delta in between the version they currently have and the version they are updating to (e.g. 1.1.18-delta, 1.1.19-delta, 1.1.20-delta will all be downloaded when updating from 1.1.17 to 1.1.20).

**Reply by caesay** [reply]
One additional note about `vpk upload`, you absolutely need to store at least a handful of updates on your server, you need more than just the latest version. If you do not do this, your users will need to download more full releases as deltas will not be available to them. `vpk upload` has a retention policy argument, which you can use to keep a predefined number of versions (e.g. 10) to limit the amount of disk space you are using on your server. For optimal experience of your users, you should retain the maximum number of releases you are comfortable with based on disk space used.
